File: python/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        sum = 0
        dic = {}
        dic[0] = 1
        for i in range(len(nums)):
            sum += nums[i]
            if sum-k in dic:
                count += dic[sum-k]
            dic[sum] = dic.get(sum, 0)+1
        return count

# Time Complexity :
#     O(N) -> Where N is the size of the array and we are iterating over the array once

# Space Complexity:
#     O(N) -> Creating a hashmap/dictionary

File: go/0560-subarray-sum-equals-k.go
func subarraySum(nums []int, k int) int {
	prefix := map[int]int{0: 1}
	curSum := 0
	count := 0
	for _, n := range nums {
		curSum += n
		if _, ok := prefix[curSum-k]; ok {
			count += prefix[curSum-k]
		}
		prefix[curSum]++
	}
	return count
}
